File: jar-shell/src/test/java/com/beancounter/shell/integ/StaticServiceTest.java
package com.beancounter.shell.integ;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.beancounter.common.exception.BusinessException;
import com.beancounter.common.model.Portfolio;
import com.beancounter.shell.config.ShareSightConfig;
import com.beancounter.shell.service.AssetService;
import com.beancounter.shell.service.PortfolioService;
import com.beancounter.shell.service.StaticService;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.cloud.contract.stubrunner.spring.AutoConfigureStubRunner;
import org.springframework.cloud.contract.stubrunner.spring.StubRunnerProperties;
import org.springframework.test.context.ActiveProfiles;

@Slf4j
@ActiveProfiles("test")
@AutoConfigureStubRunner(
    stubsMode = StubRunnerProperties.StubsMode.LOCAL,
    ids = "org.beancounter:svc-data:+:stubs:10999")
@SpringBootTest(classes = {ShareSightConfig.class})
public class StaticServiceTest {
  @Autowired
  private StaticService staticService;

  @Autowired
  private PortfolioService portfolioService;

  @Autowired
  private AssetService assetService;

  @Test
  void is_PortfolioFinders() {
    Portfolio portfolioByCode = portfolioService.getPortfolioByCode("TEST");
    Portfolio portfolioById = portfolioService.getPortfolioById("TEST");

    assertThat(portfolioByCode).isEqualToComparingFieldByField(portfolioById);
  }

  @Test
  void is_PortfolioIllegalArgumentsThrowing() {
    assertThrows(BusinessException.class, () ->
        portfolioService.getPortfolioByCode(null));
    assertThrows(BusinessException.class, () ->
        portfolioService.getPortfolioByCode("IA"));
    assertThrows(BusinessException.class, () ->
        portfolioService.getPortfolioById(null));
  }

  @Test
  void is_MarketIllegalArgumentsThrowing() {
    assertThrows(BusinessException.class, () ->
        staticService.resolveMarket(null, assetService, assetService.staticService));
    assertThrows(BusinessException.class, () ->
        staticService.resolveMarket("ERR", assetService, assetService.staticService));
  }
}

File: jar-client/src/test/java/com/beancounter/client/integ/StaticServiceTest.kt
package com.beancounter.client.integ

import com.beancounter.client.config.ClientConfig
import com.beancounter.client.services.StaticService
import com.beancounter.common.exception.BusinessException
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.ImportAutoConfiguration
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.cloud.contract.stubrunner.spring.AutoConfigureStubRunner
import org.springframework.cloud.contract.stubrunner.spring.StubRunnerProperties

/**
 * Client side access to static data - Markets & Currencies.
 */
@AutoConfigureStubRunner(stubsMode = StubRunnerProperties.StubsMode.LOCAL, ids = ["org.beancounter:svc-data:+:stubs:10999"])
@ImportAutoConfiguration(ClientConfig::class)
@SpringBootTest(classes = [ClientConfig::class])
class StaticServiceTest {
    @Autowired
    private lateinit var staticService: StaticService

    @Test
    fun is_GuardChecks() {
        assertThat(staticService.getCurrency(null)).isNull()
        assertThrows(BusinessException::class.java) { staticService.getCurrency("NOPE") }
    }

    @Test
    fun are_MarketsFound() {
        val markets = staticService.getMarkets()
        assertThat(markets).isNotNull
        assertThat(markets.data).isNotEmpty
    }

    @Test
    fun is_MarketIllegalArgumentsThrowing() {
        assertThrows(BusinessException::class.java) { staticService.getMarket("ERR") }
    }

    @Test
    fun are_CurrenciesFound() {
        val currencies = staticService.currencies
        assertThat(currencies).isNotNull
        assertThat(currencies.data).isNotEmpty
    }

    @Test
    fun is_CurrencyFound() {
        val currency = staticService.getCurrency("USD")
        assertThat(currency).isNotNull
        assertThat(currency).hasNoNullFieldsOrProperties()
    }
}
